src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsEFEMKernels.hpp

```diff
@@ -0,0 +1,412 @@
+    RAJA::ReduceMax< ReducePolicy< POLICY >, real64 > maxResidual( 0 );
+
+    forAll< POLICY >( m_fracturedElems.size(),
+                      [=] GEOSX_HOST_DEVICE ( localIndex const i )
+    {
+      localIndex k = m_fracturedElems[i];
+      typename KERNEL_TYPE::StackVariables stack;
+
+      kernelComponent.setup( k, stack );
+      for( integer q=0; q<numQuadraturePointsPerElem; ++q )
+      {
+        kernelComponent.quadraturePointKernel( k, q, stack );
+      }
+      maxResidual.max( kernelComponent.complete( k, stack ) );
+    } );
+
+    return maxResidual.get();
+  }
+  //END_kernelLauncher
+
+  /**
+   * @brief Copy global values from primary field to a local stack array.
+   * @copydoc ::geosx::finiteElement::ImplicitKernelBase::setup
+   */
+  GEOSX_HOST_DEVICE
+  GEOSX_FORCE_INLINE
+  void setup( localIndex const k,
+              StackVariables & stack ) const
+  {
+
+    localIndex const embSurfIndex = m_cellsToEmbSurfaces[k];
+
+    stack.hInv = m_surfaceArea[embSurfIndex] / m_elementVolume[k];
+    for( localIndex a=0; a<numNodesPerElem; ++a )
+    {
+      localIndex const localNodeIndex = m_elemsToNodes( k, a );
+
+      for( int i=0; i<3; ++i )
+      {
+        stack.dispEqnRowIndices[a*3+i] = m_dofNumber[localNodeIndex]+i-m_dofRankOffset;
+        stack.dispColIndices[a*3+i]    = m_dofNumber[localNodeIndex]+i;
+        stack.xLocal[ a ][ i ] = m_X[ localNodeIndex ][ i ];
+        stack.uLocal[ a*3 + i ] = m_disp[localNodeIndex][i];
+
+      }
+    }
+
+    for( int i=0; i<3; ++i )
+    {
+      // need to grab the index.
+      stack.jumpEqnRowIndices = m_wDofNumber[embSurfIndex] + i - m_dofRankOffset;
+      stack.jumpColIndices    = m_wDofNumber[embSurfIndex] + i;
+      stack.wLocal[ i ] = m_w[ m_wDofNumber[embSurfIndex] + i ];
+    }
+  }
+
+  /**
+   * @copydoc geosx::finiteElement::KernelBase::quadraturePointKernel
+   * @tparam STRESS_MODIFIER Type of optional functor to allow for the
+   * modification of stress prior to integration.
+   * @param stressModifier An optional functor to allow for the modification
+   *  of stress prior to integration.
+   */
+  template< typename STRESS_MODIFIER = NoOpFunctors >
+  GEOSX_HOST_DEVICE
+  GEOSX_FORCE_INLINE
+  void quadraturePointKernel( localIndex const k,
+                              localIndex const q,
+                              StackVariables & stack,
+                              STRESS_MODIFIER && stressModifier = NoOpFunctors{} ) const
+  {
+    real64 dNdX[ numNodesPerElem ][ 3 ];
+    real64 const detJ = m_finiteElementSpace.template getGradN< FE_TYPE >( k, q, stack.xLocal, dNdX );
+
+    constexpr int nUdof = numNodesPerElem*3;
+
+    real64 Kww_gauss[3][3], Kwu_gauss[3][nUdof], Kuw_gauss[nUdof][3];
+    real64 compMatrix[6][3], strainMatrix[6][nUdof], eqMatrix[3][6];
+    real64 matBD[nUdof][6], matED[3][6];
+
+    int Heaviside[ numNodesPerElem ];
+
+    m_constitutiveUpdate.GetStiffness( k, q, stack.constitutiveStiffness );
```

Comment:
nice, so no assumption of linear elasticity anymore?

Reply:
Nope. It should work with any constitutive model. However, I need to update the strain and the stress properly for a stress-dependent model. It's still kind of WIP and I think another clean up will be good to better embed the EFEM in the FEM library but this should be a good start.

Reply:
Yeah, small steps.  I am slowly working on the constitutive models / discretizationOps stuff, which will impact these calls anyway.